From f1cae1af962e15c75b7f6bfbd6e43ece49fe32e8 Mon Sep 17 00:00:00 2001
From: Edward Jones <ed.jones@embecosm.com>
Date: Tue, 16 Oct 2018 11:49:47 +0100
Subject: [PATCH] cgen: Exclude bits which are beyond the current end bit

If you have variable length instructions then decode-get-best-bits
could choose any bits as `best', even if they were not valid bits
for the instructions which it is trying to differentiate. When
this happened an assertion would be triggered when trying to
probe a bit beyond the mask of an instruction which was too short.

cgen/ChangeLog:

	* decode.scm (decode-get-best-bits): Remove any bits beyond
	startbit+decode-bitsize from the bits to be considered as 'best'.
---
 cgen/ChangeLog  |  5 +++++
 cgen/decode.scm | 12 ++++++------
 2 files changed, 11 insertions(+), 6 deletions(-)

diff --git a/cgen/ChangeLog b/cgen/ChangeLog
index 4741f3b..1fedf53 100644
--- a/cgen/ChangeLog
+++ b/cgen/ChangeLog
@@ -1,3 +1,8 @@
+2018-10-06  Edward Jones  <ed.jones@embecosm.com>
+
+	* decode.scm (decode-get-best-bits): Remove any bits beyond
+	startbit+decode-bitsize from the bits to be considered as 'best'.
+
 2017-04-13  Alan Modra  <amodra@gmail.com>
 
 	PR 20946
diff --git a/cgen/decode.scm b/cgen/decode.scm
index 615e2f8..f84fe19 100644
--- a/cgen/decode.scm
+++ b/cgen/decode.scm
@@ -394,17 +394,17 @@
 ; Also, see preceding FIXME: We can't proceed past startbit + decode-bitsize
 ; until we've processed all bits up to startbit + decode-bitsize.
 
-(define (decode-get-best-bits insn-list already-used startbit max decode-bitsize lsb0?)
+(define (decode-get-best-bits insn-list already-used startbit maxn decode-bitsize lsb0?)
   (let* ((raw-population (/distinguishing-bit-population (map insn-base-mask insn-list)
 							 (map insn-base-mask-length insn-list)
 							 (map insn-value insn-list)
 							 lsb0?))
-	 ;; (undecoded (if lsb0?
-	 ;; 		(/range2 startbit (+ startbit decode-bitsize))
-	 ;;		(/range2 (- startbit decode-bitsize) startbit)))
-	 (used+undecoded already-used) ; (append already-used undecoded))
+	 ; undecoded are all of the bits beyond start-bit + decode-bitsize
+	 (max-base-mask-len (apply max (map insn-base-mask-length insn-list)))
+	 (undecoded (/range2 (+ startbit decode-bitsize) (- max-base-mask-len (+ startbit decode-bitsize))))
+	 (used+undecoded (append already-used undecoded))
 	 (filtered-population (/vector-copy-set-all raw-population used+undecoded #f))
-	 (favorite-indices (/population-top-few filtered-population max))
+	 (favorite-indices (/population-top-few filtered-population maxn))
 	 (sorted-indices (sort favorite-indices (lambda (a b) 
 						  (if lsb0? (> a b) (< a b))))))
     (logit 3
-- 
2.9.5

